File: 01-Leetcode/647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:

    def countSubstrings(self, s: str) -> int:
        n = len(s)
        cnt = [0]

        def checkPd(il, ir):
            while il >= 0 and ir < n and s[il] == s[ir]:
                cnt[0] += 1
                il -= 1
                ir += 1

        for i in range(n):
            # odd len
            checkPd(i, i)

            # even len
            checkPd(i, i + 1)
        return cnt[0]


s = Solution()
assert s.countSubstrings("aaa") == 6
assert s.countSubstrings("abc") == 3

# @lc code=end

File: 01 - Leetcode/647.palindromic-substrings.cpp
/*
 * @lc app=leetcode id=647 lang=cpp
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
class Solution
{
public:
    int cnt = 0;
    void checkLpd(int il, int ir, int lenPd, string s)
    {
        int n = s.size();
        while (il >= 0 && ir < n && s[il] == s[ir]) // O(n)
        {
            cnt += 1;

            il--;
            ir++;
        }
    }

    int countSubstrings(string s)
    {
        int n = s.size();

        // O(n^2)
        for (int i = 0; i < n; i++) // O(n)
        {
            // odd length
            checkLpd(i, i, 1, s);

            // even length
            checkLpd(i, i + 1, 1, s);
        }

        return cnt;
    }
};
// @lc code=end
